webservice/components/request_adapter.py

```diff
@@ -70,3 +75,106 @@ def _get_url(self, url=None, url_params=None, **kwargs):
             url = self.collection.url
         url_params = url_params or kwargs
         return url.format(**url_params)
+
+
+class OAuth2RestRequestsAdapter(Component):
+    _name = "oauth2.requests"
+    _webservice_protocol = "http+oauth2"
+    _inherit = "base.requests"
+
+    def __init__(self, *args, **kw):
+        super().__init__(*args, **kw)
+        # cached value to avoid hitting the database each time we need the token
+        self._token = {}
+
+    def fetch_token(self):
+        pass
+
+    def _is_token_valid(self, token):
+        """We consider that a token in valid if it has at least 10% of
+        its valid duration. So if a token has a validity of 1h, we will
+        renew it if we try to use it 6 minutes before its expiration date.
+        """
+        expires_at = token.get("expires_at", 0)
+        expires_in = token.get("expires_in", 3600)  # default to 1h
+        now = time.time()
+        return now <= (expires_at - 0.1 * expires_in)
+
+    @property
+    def token(self):
+        """Return a valid oauth2 token.
+
+        The tokens are stored in the database, and we check if they are still
+        valid, and renew them if needed.
+        """
+        if self._is_token_valid(self._token):
+            return self._token
+        backend = self.collection
+        with backend.env.registry.cursor() as cr:
+            if backend.env.context.get("test_mode"):
```

Comment:
why this hack?

Reply:
Originally, to avoid a commit during the tests. I'll check if it is really needed.

Reply:
issue in the tests where the backend record is not flushed to disk and cannot be found by the SELECT, if I remember correctly. I'm removing this to see if I called `flush_model()` correctly at the time.

Reply:
https://github.com/OCA/web-api/actions/runs/8322619666/job/22770728397#step:8:204 is what I get.